lib/stats/stats-cluster.c

```diff
@@ -28,15 +28,25 @@
 #include <string.h>
 
 GPtrArray *stats_types;
-static const gchar *module_names[];
 
 void
 stats_cluster_init(void)
 {
   g_assert(!stats_types);
   stats_types = g_ptr_array_new_with_free_func(g_free);
-  for (int i = 0; i < SCS_MAX; i++)
-    g_ptr_array_add(stats_types, g_strdup(module_names[i]));
+
+  g_assert(stats_register_type("none") == 0);
+  g_assert(stats_register_type("group") == SCS_GROUP);
+  g_assert(stats_register_type("global") == SCS_GLOBAL);
+  g_assert(stats_register_type("center") == SCS_CENTER);
+  g_assert(stats_register_type("host") == SCS_HOST);
+  g_assert(stats_register_type("sender") == SCS_SENDER);
+  g_assert(stats_register_type("program") == SCS_PROGRAM);
+  g_assert(stats_register_type("severity") == SCS_SEVERITY);
+  g_assert(stats_register_type("facility") == SCS_FACILITY);
+  g_assert(stats_register_type("tag") == SCS_TAG);
+  g_assert(stats_register_type("filter") == SCS_FILTER);
+  g_assert(stats_register_type("parser") == SCS_PARSER);
```

Comment:
Why are these asserts? (or, why are these asserts are missing in other cases?).

What if we would keep the array (a "core" array with these names).

Reply:
When we had a separate array for the names, it had a certain risk: the list of names and the enum could diverge. This happened once with `SCS_STDIN`, and was fixed in 5a071f4eac89aace93046d06b5a0f532b9767a42. That's when we added a static assert to check if the size of the two entities are the same. Checking lengths mitigates some of the risks, but could not catch bugs coming from when the items in the of them are reordered while the other one did not follow this change up. I think this kind of bug has not yet happened, so maybe I am too paranoid here.

Bringing back the original array would result in little more complex code, at least to my taste: we need a static array, we need a dynamic array, we need to test SCS_STATIC_MAX with the static array length (I would not let this check as we had run into problem due to this). Then the static array still need to be registered, because the dynamic registration should not overlap with it. This latter problem could be resolved in other way, for example dynamic allocation would return dynamic_value + SCS_STATIC_MAX, and during counter registration, we could decide if the counter is static or dynamic by comparing the input value with SCS_STATIC_MAX, and thus branch the lookup in the static array or in the dynamic array. Me feels this version simpler.

I can drop the g_asserts though if you think we do not have much benefit from them. Then I would just g_assert the lengths.

Reply:
okay, and can't we simply set these values during registration (I mean do we need these pre-defined SCS_xxx values)?

Reply:
Good idea. I can make these availables via extern. I will do this.

Reply:
Instead of extern, I created a stateful function that is responsible for getting and setting these constants. This has two benefits: these "constants" are not mutable through accidental assignment, and also I can assert if caller tries to use them before initialization.

Reply:
My new version was not very good either. For now I dropped this patch. Lets address this at a later time.